### ISSUE TITLE
False positive for `used-before-assignment` in `else` clause when the only non-`break` exit from a loop is an except block

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#5683]
### Bug description

```python
def do_something():
    pass

for _ in range(3):
    try:
        do_something()  # not an exit branch because of `else` below (will only exit here with an exception)
    except ZeroDivisionError as e:
        error = e  # < exit branch where error is defined
    else:
        break  # < exit condition where error is *not* defined, but will skip `raise error` below
    print("retrying...")
else:
    raise error
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:13:10: E0601: Using variable 'error' before assignment (used-before-assignment)
```

### Expected behavior

No message. Identified by @martimlobao and discussed in detail https://github.com/PyCQA/pylint/pull/5402#issuecomment-1012791780.

Quoting from my [suggested](https://github.com/PyCQA/pylint/pull/5402#issuecomment-1013209351) solution:

So maybe this is the constraint to check for:

- if the use of `error` being tested is guarded by `else` in `for/else` (or `while/else`)
- if the definition of `error` in an `except` block is a descendant of the same `for/while`
- if that inner `else` breaks
- if there is no `continue` anywhere at all underneath the loop and before the break (other than inside another loop, but including nested conditions and try/excepts)

### Pylint version

```shell
pylint 2.13.0-dev0
astroid 2.9.2
Python 3.10.1 (v3.10.1:2cd268a3a9, Dec  6 2021, 14:28:59) [Clang 13.0.0 (clang-1300.0.29.3)]
```
